### PR TITLE
big refactoring, first step in merging VNC and CLI versions of the code

### PR DESCRIPTION
**Pull request by mrosseel**
NOTE: VNC GUI version is not working currently

What was done:
* refactoring of all handpad interactions into its own class
* if the '-fh' or fake handpad option is passed, the handpad output is shown in the console and can be navigated with the keyboard
* created EFinder class containing all relevant methods
* added argparse CLI interface to eFinder.py
* extracted all GUI related methods to gui.py, still to be investigated how to further merge this

### CODE REVIEW

**Comment by mrosseel** [comment]
@WimDeMeester just added the PR for the refactoring. To be clear, this shouldn't be merged.

It would however help if you could test it, but *not* under a clear sky because there are surely plenty of bugs.
I've tested as far as I could but of course goto's and real nexus interactions are impossible to test.

Next task is integrating the GUI again, but I feel most of the hairy parts are done.

**Comment by WimDeMeester** [comment]
> It would however help if you could test it, but not under a clear sky because there are surely plenty of bugs.

I'll try to do this tomorrow evening!

**Comment by mrosseel** [comment]
> > It would however help if you could test it, but not under a clear sky because there are surely plenty of bugs.
> 
> I'll try to do this tomorrow evening!

great, ui is also kinda working (add the -g flag) but many missing pieces there. 
The non-ui version should be better to test. If you don't want to use the handpad you can use the arrow keys, enter and space for navigating the handpad, whose current state is shown in the console.

**Comment by WimDeMeester** [comment]
Tested it very quickly in a Virtual Machine.  The poetry update worked fine, and I can start up `eFinder.py`.

**Comment by WimDeMeester** [comment]
Tested and some problems to start up 😛 
This is the log:
[errors.log](https://github.com/WimDeMeester/eFinder/files/10297253/errors.log)

**Comment by mrosseel** [comment]
seems like the nexus couldn't be found on port /dev/ttyS0 - but it could also be triggered by several incorrect logging statements in nexus.py. That's one of the files I can't test locally due to no nexus :)

Could you try again?

**Comment by WimDeMeester** [comment]
Progress!  The new log:
[errors.log](https://github.com/WimDeMeester/eFinder/files/10298473/errors.log)

**Comment by mrosseel** [comment]
> Progress! The new log: [errors.log](https://github.com/WimDeMeester/eFinder/files/10298473/errors.log)

wgs84 should be fixed now, along with another bug that was lurking in the nexus code

**Comment by WimDeMeester** [comment]
Testing will be for tomorrow, as we are leaving to my parents for the Christmas celebrations.
Merry Christmas to all of you, @AstroKeith and @mrosseel

**Comment by WimDeMeester** [comment]
New test, new error log:
[errors.log](https://github.com/WimDeMeester/eFinder/files/10300531/errors.log)

**Comment by WimDeMeester** [comment]
Tested it again, no errors anymore.  
I first started eFinder.py (so without the GUI).  Everything started correctly, the camera and the Nexus were found, but the handpad does not work...
Second test was to start the GUI version (using the -g option).  Starts fine, but also here, the handpad does not work.  I tried to take a fake image (of M 31 and of polaris), but I got a white image...

**Comment by mrosseel** [comment]
@WimDeMeester did some more work on the Gui, also extracted connection type from Nexus in its own classes, and made Nexus inherit from NexusInterface so the debug class can be swapped without python linter warnings.

About the handpad: I don't have one, and can't see anything wrong in the code. Could you run with the -x flag and see if there are errors? If not, check that the port is correct for the handpad, that it still works in old code etc. Might be something stupid but can't recreate. 
It's indeed correct that the arrow key navigation of the handpad only works with the -fh flag, because if you have a real one this should work :)

In short: I'm progressing, if you could do some more handpad testing and we fix that, we can move on to the correct functioning of gui-less and gui operation

**Comment by mrosseel** [comment]
update: found the handpad misunderstanding, there's indeed a conflict between regular handpad and handpad usage in UI, will think on how to solve it. GUI is progressing with capture, solve, align, anotate and display mostly working, todo is the measure operations and goto via align, goto via move. Most valuable would be non-gui testing, since most of the code is the same anyway.

**Comment by WimDeMeester** [comment]
I don't have a lot of time now, but I did some tests.  
I tried the handpad version, but I still can not use the handpad at all...  No reaction when I use the button.  I'll try to check this this evening.
I also tried the gui version.  I selected polaris from the option.  I expect to get the polaris picture shown, but I get a new picture taken from the camera.  This should not be the case...

**Comment by mrosseel** [comment]
Test image was fixed, only worked with fake camera.
Known issues:
- exp range and gain range not getting loaded/shown correctly
- Goto not working

**Comment by WimDeMeester** [comment]
Very quick test.  Using the example images now works without problems.  When I show the image and I then select the 'Use current value' for the Polaris offset, there is a small GUI glitch:

image

I tried to measure the offset from the polaris test image and I get the following error:
Screenshot 2023-01-02 at 13 42 39

**Comment by WimDeMeester** [comment]
I have fixed the problem with the handpad.  We were first comparing strings with Enum values.  I corrected that, so everything seems to work.
Some things I found while playing with the handpad:
- In the previous version, when you are in the start menu, you can move to the right to see the Nexus location and the other menus.  Now, I first have to move to the left, before I can see the Nexus location (and then to the right for the rest). See https://github.com/WimDeMeester/eFinder#in-the-handpad-version
- When trying to set the exposure (by moving down, and then right, then up to move the value), I get the following error message: 
image

I can have a look at these things, but it will be for later this week.

**Comment by WimDeMeester** [comment]
Fixed the problem with the fake and real handpad.  I reverted my changes and made sure that the fake handpad uses the same code as before.  For the real handpad, I now send the DisplayButtons enum instead of a string.

**Comment by WimDeMeester** [comment]
@mrosseel , I will let you fix the problem with the up_down_inc() and flip().  I think we are missing an attribute parameter, but I don't know what this should be...

**Comment by mrosseel** [comment]
> @mrosseel , I will let you fix the problem with the up_down_inc() and flip(). I think we are missing an attribute parameter, but I don't know what this should be...

I have already fixed the attribute param error locally.
Thx for the handpad fix - still one issue preventing me from checking in, should be done tomorrow

**Comment by WimDeMeester** [comment]
I tried the new version.  I first updated the poetry packages.  I get the following error when using a fake Nexus (but the real camera and the real handpad), starting up the GUI version:
image

When I start the non-GUI version (still with a fake nexus), and I go to the settings pages, and then to exposure, I get the following error:
image

Last test I did was with the non-GUI version, with fake nexus and fake handpad.  I get the following error when I push 'Down', 'Right', 'Up':
image

**Comment by mrosseel** [comment]
Sorry should have notified you earlier, it works locally but haven't
checked in yet

**Comment by WimDeMeester** [comment]
Sorry for the misunderstanding

**Comment by WimDeMeester** [comment]
Hi @mrosseel
Did you had some time to work more on this?  Your last messages says that you have something working locally, but that you didn't check it in yet...

**Comment by mrosseel** [comment]
Hi,

Lots of other work + blocked on the config file which still had some issues
(working but not good enough),  should have a bit more time next week

**Comment by WimDeMeester** [comment]
Hi @mrosseel, 

Any status update on this?  If you don't have the time at this moment, can you check in the latest version of the source code and write a list with things that still need to be done?  I'll have a look at it then.

**Comment by mrosseel** [comment]
some progress, still thinking about config stuff, maybe yaml/json would be
better/more standard.
Or reverting most of the changes...
Will change the config values back to what they were and check in soon.

Mike

**Comment by WimDeMeester** [comment]
Thanks, @mrosseel !
I think yaml would be a good option for the config files.

**Comment by mrosseel** [comment]
yaml configuration was added and I committed the code. Need to do more
testing to make sure it works correctly, but you can already take a
look  if you want.

**Comment by mrosseel** [comment]
what I know now:

- config might be still broken in some places
- thread complaint about the gui loop, should be fixed but gui seems to work
- making sure all gui stuff is extracted and deleting the VNC class

**Comment by mrosseel** [comment]
I'll fix some more config stuff and see if I can recreate this.
Feel free to test again tonight and fix some things, but don't forget to
update.

Mike

**Comment by mrosseel** [comment]
pushed an update, config should work now (eyepieces), having some trouble
with the sidereal update thread which is not working atm.
Haven't retested your scenarios

Mike

**Comment by WimDeMeester** [comment]
Just tested the latest version.  The eyepieces are now indeed visible and can be used.  I think that the other problems are still in the software.  
I found another (cosmetic) problem.  When starting up the GUI, the exposure and gain are a bit weird.
image

I'll start to investigate the problems one by one...

**Comment by WimDeMeester** [comment]
Now the coordinates of the solved field are displayed.
image
The delta x,y are not correct yet.

**Comment by WimDeMeester** [comment]
Found another small problem in the GUI:

- When clicking on the 'Used Saved' option for the offset, a part of the image is black.
image

**Comment by mrosseel** [comment]
@WimDeMeester got sidetracked by a GUI threading issue which was due to the fact I was running the UI through  'ssh -Y', and it doesn't happen on the pi itself 🙄 

- cannot reproduce the handpad bug
- added a '.editorconfig' file because our code formatting differs and that makes it hard when merging. Visual studio code has a plugin for this, neovim >0.9 does it automatically. Contents of the file is just defaults, feel free to change.

I'm checking the 'measure' bug now

**Comment by WimDeMeester** [comment]
@mrosseel 
I tested the measure bug, but it is still not working for me...  I have the impression that when using the real camera, that no image is used, even when the polaris image is selected...  
Anyway, as far as I remember, the image should be shown when clicking the measure button.  This is not the case at the moment.

**Comment by mrosseel** [comment]
> @mrosseel I tested the measure bug, but it is still not working for me... I have 

I fixed a crash, what you are describing is a new bug related to the measure button. Will retest, can you also check the handpad because I can't make it crash (I'm using the fake handpad of course). The exposure and gain radiobuttons have resisted my attempts to fix them up till now.
Will have another look tomorrow.

**Comment by WimDeMeester** [comment]
For the handpad bug, I print out line1 and the type of line1.  For the real handpad, I get 
image
when I go down and once to the right.  
For the fake handpad, the code in line 70 of Display.py is not executed.  
I can check if line1 is an int, and if that is the case, I can convert it to a str.  I need to do the same for a float then.  I don't know if this is the best solution...

**Comment by WimDeMeester** [comment]
I think I fixed the problem with the crashing handpad in 0372283.  However, the values are not changed at all...  I don't see how I can fix this.  Can you have a look, @mrosseel ?

**Comment by mrosseel** [comment]
You mean the values don't change on the UI, the config file, thebhandpad or
all of the above?

**Comment by WimDeMeester** [comment]
In any case not on the handpad... I did not check the file

**Comment by WimDeMeester** [comment]
> In any case not on the handpad... I did not check the file

The file and the GUI are also not adapted.

**Comment by WimDeMeester** [comment]
The problem with the handpad is that the changes are not displayed on the handpad.  If I change the value of the exposure, the same value is shown on the handpad.  If I go back to the overview page, I can see the updated value.

**Comment by WimDeMeester** [comment]
Hi @mrosseel,

Is there still something you want to change?  If not, I will accept this pull request and start working on the main branch to fix the last bugs.

**Comment by mrosseel** [comment]
hi, just committed what should be the fix for the handpad, could only test in ascii output mode of course. Ok to merge this, let me know if there are any other issues you notice.

**Comment by WimDeMeester** [comment]
Thanks,
I will probably test this evening with the real handpad.  
A small request:  Can you make sure that you cannot end up with negative exposure times?

**Comment by WimDeMeester** [comment]
I get this error when I start up the GUI (with real hardware).  I don't know if this is a problem or not...
image

**Comment by WimDeMeester** [comment]
The real handpad seems to work fine!
But I get the following error when using the real camera and when I click on 'Measure': 
image

**Comment by mrosseel** [comment]
sure I'll add that

**Comment by mrosseel** [comment]
fixed the negative issue (I did it for al increments/decrements so gain is also always 0 or more)
The crash was just a bad default return value when the solve fails.

**Comment by WimDeMeester** [comment]
I needed to reinstall my raspberry pi installation.  I used 64 bit this time, but I don't know if this is relevant for this problem...
When I select 'Polaris' and press 'Measure', I now get the following result:
image
In the past, the values were 14.2, 12.2, now I see -160, 120...

**Comment by AstroKeith** [comment]
Thats strange!

Although in fact the 64bit OS for Pi isn’t recommended. Way back I tried the basic solver code on a 64bit build, and couldnt see any speed improvements. Speaking with the astronomy.net <http://astronomy.net/> author he wasnt surprised as the limiting factor is memory access speed.

BTW, I made a small change to my version recently. 
The ‘home’ screen now shows the live Nexus RA & Dec values, updated at 2 Hz.
From this Home Screen, a short select press does a ‘local sync’ (hence re-calibrating that live RA & Dec display)
A long press does a repeat last goto, (not a goto++)

One of my users asked for this, and nearly everyone else has installed it too. It’s very useful for hunting a minor planet, comet or supernova, etc when only the RA & Dec is known.

I’ve used it to recover position after viewing a goto target, moving the scope to look at an associated feature, and then recover the original position. I observe Galaxy Cluster quite a lot and the catalogue RA & Dec is only the brightest member of the group, so manual search is needed.

I’ve also started to add bits to the homescreen.

On eFinder only systems, the stop right hand corner has a “N” for not Nexus aligned, ’T’ for aligned.
On ScopeDog mk3’s as well as this, the bottom right corner has a '1,2 or 3’ to show current selected joystick slew speed.

I think our codes versions are so diverged, my actual code isn’t much use to you?! Happy to share though if wanted.

The basic eFinder handsethas got a spruce up too! The 5 way switch is expensive, but very nice. Photo below,

Cheers
Keith

￼

**Comment by WimDeMeester** [comment]
Hi Keith,

Solving does work when I just take the example image of polaris or M31 in the eFinder GUI, and then push solve.  The correct information is returned.  I guess there is an error somewhere in our code.

The picture did not come through.  Can you try to upload it again?  We are interested in the changed code if you want to share it!

**Comment by AstroKeith** [comment]
IMG_5593

**Comment by WimDeMeester** [comment]
> IMG_5593

Looks indeed much more professional!

**Comment by WimDeMeester** [comment]
> 



> I needed to reinstall my raspberry pi installation. I used 64 bit this time, but I don't know if this is relevant for this problem... When I select 'Polaris' and press 'Measure', I now get the following result: image In the past, the values were 14.2, 12.2, now I see -160, 120...

There was a problem with the latest version of astrometry.net.  When I reverted from version 0.94 to 0.93, the results are correct again.

**Comment by WimDeMeester** [comment]
Tested with real hardware today.  I still have the following problems:
- [ ] Cannot access the handpad anymore: `ls: cannot access '/dev/serial/by-id/'`
- [ ] ASI driver does not work.  ELF32 error.  This has to do with the 64 bit installation.  I will have a look the next days on how to solve this.

**Comment by WimDeMeester** [comment]
I did some tests and I get the following problems:
- [ ] When the solving fails (because I took an image during the day), I get the following error: 
image
- [ ] When I try to do an align, when nothing was solved, I get the following error: 
image

**Comment by AstroKeith** [comment]
Hi Wim & Mike,

I’ve been toying with Skyfield and what it can do. This coincided with someone wanting bespoke finder charts for comets and minor planets. Easy.

Well it would be easy, except the database from the MPC requires you to know the designation in its exact form, to search for it.  I’m still trying to develop a better search tool.

But then I thought I could add a comet/minor planet overlay on the eFinder GUI image. It’s not too difficult I think. Would it be useful?

Developing a method for selecting the right object is the most difficult. Especially using tkinter!

What do you think?

Keith

**Comment by AstroKeith** [comment]
Well spotted,

My prototype stored the comet data file, and checked on boot to see if internet was available to update it.

This may be OK for Comets as the data isnt changed that often. It’s a simple matter of plugging the Pi into your home network (via ethernet) on boot to get an update though.

Asteroids is a bigger problem. The data continually changes and there are millions of them. I could find a way of filtering out the dross, but it’s probably that if someone was chasing an asteroid, it would be a new or strange one.

On my bucket list is the task of writing an Android App to replace the efinderVNCGUI.py functionally. Perhaps I should wait until I have done that ( a year or two!)

Keith

**Comment by WimDeMeester** [comment]
Good solution to check for an internet account.  An Android app would indeed also solve a lot of problems!

**Comment by AstroKeith** [comment]
Couldnt resist taking a look at the Android App challenge.

Turns out it’s very similar to Java, with the complications mostly in the UI. I’m running Android Studio and its pretty impressive - including a pretty good device emulator. After a day of learning I’m getting my own basic Apps running.

The big challenge will be the communication with the Pi. Lots of options and none simple. So some questions for you…

What would be your preferred connection method?

USB
Bluetooth (slow)
Wifi peer to peer
Wifi Access Point (ie Pi joins the Device, or vice versa
Wifi infrastructure mode - use home router, portable router or Nexus DSC as a router.

Your choice may well be determined for you by your other connections. I’m using USB from Pi to Nexus DSC, Nexus DSC wifi to SkySafari, so I have an unused Pi wifi. I can also use the Nexus DSC as a wifi router but likely to get problems running SkySafari in parallel with my new App down the same device wifi connection. The Nexus DSC Pro cant act as a router.

Rather than try and reproduce the VNCGUI screen on the new App, I’m thinking of making an optional ‘plug-in’ to the hand box eFinder version. It would be just a basic display of the captured image (useful for focusing and parameter tweaking), and a display of an annotated image. Thus the ‘buttons’ on the app would be just two - for load last image or capture new image, and a series of check boxes to decide the annotation layers (including none).

Any though

**Comment by AstroKeith** [comment]
Having spent a weekend indoors hiding from the heatwave, I can see that establishing comms between an Android device and the Pi is going to be a challenge and a half! Exchanging messages is foreseeable but transmitting whole images might take some real learning!

**Comment by WimDeMeester** [comment]
Hi @AstroKeith , I have started a discussion on an Android App.  You can see it here: #54

**Comment by WimDeMeester** [comment]
> I did some tests and I get the following problems:
> 
> * [ ]  When the solving fails (because I took an image during the day), I get the following error:
>   image
> * [ ]  When I try to do an align, when nothing was solved, I get the following error:
>   image

Hi @mrosseel , 
Did you already find some time to look at the last things to fix / implement?

**Comment by WimDeMeester** [comment]
Hi @mrosseel , these two errors are fixed.  I now get a similar error in line 326 of gui.py.
